compiler/noirc_frontend/src/hir_def/types/arithmetic.rs

```diff
@@ -111,6 +111,10 @@ impl Type {
 
                 Type::CheckedCast { from: Box::new(from), to: Box::new(to) }
             }
+            Type::Array(length, element) => Type::Array(
+                Box::new(length.canonicalize_helper(found_checked_cast, run_simplifications)),
+                element,
+            ),
```

Comment:
I can see what this does but can you elaborate why it was added? The intent of canonicalize was to only work on numeric types. If we start trying to find and canonicalize types recursively like this we're missing a bunch of cases, e.g. numerics in struct generics, handling type variables, etc.

Reply:
I can see this is blocking a sync but I still want to ensure its the right solution before just merging it